encoding: avoid heap allocation when encoding/decoding SpatialObject

nvb opened the pull request:
This commit shaves off a heap allocation in each of the following functions:
- `EncodeTableKey`
- `EncodeGeoAscending`
- `EncodeGeoDescending`
- `DecodeGeoAscending`
- `DecodeGeoDescending`
- `EncodeGeoValue`
- `EncodeUntaggedGeoValue`
- `DecodeUntaggedGeoValue`

In each of these, we were letting a local SpatialObject variable escape
to the heap when passed through `protoutil.Marshal` or `protoutil.Unmarshal`.
This was unnecessary, as we already have a Datum on the heap nearby any
of the callers of these functions, so with a bit of restructuring, we
can avoid the allocations entirely.

The callers of the decode functions are a little more awkward than
strictly necessary right now. They are written the way that they are so
that we don't accidentally regress on this improvement when we remove
the double-boxing being discussed in #53252.

I don't have any benchmarks set up, so I wasn't able to measure the
exact effect of this change.

nvb committed:
encoding: avoid heap allocation when encoding/decoding SpatialObject

This commit shaves off a heap allocation in each of the following functions:
- `EncodeTableKey`
- `EncodeGeoAscending`
- `EncodeGeoDescending`
- `DecodeGeoAscending`
- `DecodeGeoDescending`
- `EncodeGeoValue`
- `EncodeUntaggedGeoValue`
- `DecodeUntaggedGeoValue`

In each of these, we were letting a local SpatialObject variable escape
to the heap when passed through protoutil.Marshal or protoutil.Unmarshal.
This was unnecessary, as we already have a Datum on the heap nearby any
of the callers of these functions, so with a bit of restructuring, we
can avoid the allocations entirely.

The callers of the decode functions are a little more awkward than
strictly necessary right now. They are written the way that they are so
that we don't accidentally regress on this improvement when we remove
the double-boxing being discussed in #53252.

I don't have any benchmarks set up, so I wasn't able to measure the
exact effect of this change.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53304)
<!-- Reviewable:end -->

nvb committed:
geo: avoid two heap allocations in compareSpatialObjectBytes

protoutil.Marshal causes both arguments to escape to the heap.
This is easily avoided.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

(I'm starting to yearn for explicit heap allocations instead of this Go "magic")

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53304#-:-MFVLltm0uN7DMWwyBpR:b-gmm27r)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
> (I'm starting to yearn for explicit heap allocations instead of this Go "magic")

Agreed. The general rule of thumb that helps me is that interfaces == heap memory, so either a) avoid them, b) make sure they're inlined away, c) make sure they point to existing heap memory.

bors r+
